### PR TITLE
Fix: fencer: broadcast returned fencing operations and update outdated pending ones in the history

### PR DESCRIPTION
**Pull request by gao-yan**
If a fencer had a pending fencing operation in the history but somehow
was not reachable when the fencing operation returned, the pending
operation would remain outdated and get stuck in the history.

The solution here is to broadcast returned fencing operations and update
outdated pending ones in the history.

### CODE REVIEW

File: daemons/fenced/fenced_history.c [hunk 1]
```diff
@@ -363,7 +363,18 @@ stonith_merge_in_history_list(GHashTable *history)
...
+                && stored_op->state != st_done
+                && (op->state == st_failed || op->state == st_done)) {
+                    crm_debug("Updating outdated pending operation %.8s "
+                              "(state=%s) according to the one (state=%s) from "
+                              "remote peer history",
+                               op->id, stonith_op_state_str(stored_op->state),
+                               stonith_op_state_str(op->state));
+
+            } else {
+                continue; // Skip existent (@TODO state-merging might be desirable)
```
Comment (kgaillot) [inline daemons/fenced/fenced_history.c:376]: I think you can drop the TODO, you did it :)

**Review by kgaillot — COMMENTED** [review]
Looks good to me. @wenningerk , do you want to take a look too?

File: daemons/fenced/fenced_history.c [hunk 2]
```diff
@@ -288,9 +288,26 @@ stonith_local_history_diff(GHashTable *remote_history,
...
+                        g_hash_table_lookup(remote_history, op->id);
+
+                    if (remote_op) {
+                        if (remote_op->state != st_failed
+                            && remote_op->state != st_done
+                            && (op->state == st_failed
+                                || op->state == st_done)) {
+
+                            crm_debug("Broadcasting operation %.8s (state=%s) to "
+                                       "update the outdated pending one "
```
Comment (wenningerk) [inline daemons/fenced/fenced_history.c:302]: not sure what our current guidelines say but the rest in the code in this file aligns independently of the tab-space in these cases.

**Comment by wenningerk** [comment]
Looks like the natural thing to do.
A bit disturbing that it hadn't been done like that - by me presumably - before ;-)

What we could do is concentrate the new logic in one place + improve performance
if we in the diff-loop as well check for if the remote entry should overwrite our local
one and for that case remove the entry from the local hash-table otherwise we
remove the matching entry from the remote-hash-table.
The merge would degrade to unconditionally moving everything from the 
remote-hash-table to the local one.

**Comment by wenningerk** [comment]
Actually off topic here but if nobody has had a look yet CI failing is due to new false positives in date evaluation with updated coverity:

16:48:54 Error: OVERFLOW_BEFORE_WIDEN (CWE-190):
16:48:54 /srv/workspace/pacemaker-build-covscan/pacemaker-build-covscan/coverity/lib/common/iso8601.c:337: overflow_before_widen: Potentially overflowing expression "2592000 * utc->months" with type "int" (32 bits, signed) is evaluated using 32-bit arithmetic, and then used in a context that expects an expression of type "long long" (64 bits, signed).
16:48:54 /srv/workspace/pacemaker-build-covscan/pacemaker-build-covscan/coverity/lib/common/iso8601.c:337: remediation: To avoid overflow, cast either "2592000" or "utc->months" to type "long long".
16:48:54 
16:48:54 Error: OVERFLOW_BEFORE_WIDEN (CWE-190):
16:48:54 /srv/workspace/pacemaker-build-covscan/pacemaker-build-covscan/coverity/lib/common/iso8601.c:341: overflow_before_widen: Potentially overflowing expression "86400 * (utc->days - 1)" with type "int" (32 bits, signed) is evaluated using 32-bit arithmetic, and then used in a context that expects an expression of type "long long" (64 bits, signed).
16:48:54 /srv/workspace/pacemaker-build-covscan/pacemaker-build-covscan/coverity/lib/common/iso8601.c:341: remediation: To avoid overflow, cast either "86400" or "utc->days - 1" to type "long long".

**Comment by wenningerk** [comment]
> The merge would degrade to unconditionally moving everything from the
> remote-hash-table to the local one.

Not quite true. We would still need to check for pending actions that are
said to be handled by the local node but have no representation there.

File: daemons/fenced/fenced_history.c (hunk 1 above)
Comment (gao-yan) [reply daemons/fenced/fenced_history.c]: Right :-)

**Comment by gao-yan** [comment]
> > The merge would degrade to unconditionally moving everything from the
> > remote-hash-table to the local one.
> 
> Not quite true. We would still need to check for pending actions that are
> said to be handled by the local node but have no representation there.

Unless I've missed anything, this has been handled already, right?

In stonith_local_history_diff(), as long as any op existing in the local doesn't exist in the remote, it'll be added to be broadcasted. And in stonith_merge_in_history_list(), as long as any op existing in remote doesn't exist in the local, it'll be added into the local.

**Comment by wenningerk** [comment]
> > > The merge would degrade to unconditionally moving everything from the
> > > remote-hash-table to the local one.
> > 
> > 
> > Not quite true. We would still need to check for pending actions that are
> > said to be handled by the local node but have no representation there.
> 
> Unless I've missed anything, this has been handled already, right?
> 
> In stonith_local_history_diff(), as long as any op existing in the local doesn't exist in the remote, it'll be added to be broadcasted. And in stonith_merge_in_history_list(), as long as any op existing in remote doesn't exist in the local, it'll be added into the local.

And if it is a pending action that is claimed to be executed by the local node but this one hasn't any records (e.g. fencer or whole node rebooted) it is already failed.
Just had said if we are removing from the two hashes during diff-creation we could blindly move what stayed in the remote-hash to the local one. But the failing-logic has to stay.

**Comment by gao-yan** [comment]
There might be misunderstanding here.

The newly added logic works like:

• It takes effect only if any op exists both in the local and in the remote, no matter on broadcasting or merging.

• On broadcasting, rather than reducing anything from the diff, it only additionally adds any op into the diff if it's shown being finished in the local but being pending in the remote.

• On merging, overriding of the state in the local is allowed to proceed additionally by this logic only one way ( "finished" overrides "pending"): if any op is shown being finished in the remote but being pending in the local.

So the failing logic still works if any op is shown being pending in the remote but doesn't exist in the local. Of course if any op is shown being finished in the remote but being pending in the local,  we'll have respect the "finished" one.

**Comment by wenningerk** [comment]
> There might be misunderstanding here.
> 
Yes, I got how that works.
We've just added additional checking code here and I was trying to think over how we could improve efficiency - probably more than making up for the newly introduced checking.
And in my first suggestion I had overlooked the case where these actions are artificially failed.
So I commented that we would have to keep that part if we do the modification along the lines I had mentioned.
I never said your current state wouldn't work.
But we can of course merge it as it is and think about a performance improvement afterwards.
Just thought it might even be easier to follow if the new logic is in just one place and we wouldn't add new code and move that around with the next commit.

**Comment by kgaillot** [comment]
> Actually off topic here but if nobody has had a look yet CI failing is due to new false positives in date evaluation with updated coverity:

Yes, it is indeed a false positive -- the integer arithmetic can't overflow because the days and seconds variables are bounded to low values. I have a commit that makes coverity happy but it's in a branch that isn't quite ready for PR yet overall ...

**Comment by kgaillot** [comment]
> But we can of course merge it as it is and think about a performance improvement afterwards.

Sounds good to me

**Comment by gao-yan** [comment]
Do you mean on creation of the diff, when iterating over the local hash table,  we start dropping unneeded entries from the remote hash table and even the local hash table already? Well, of course when merging the state, we'll still have to iterate over the remote hash table to go through the failing-pending logic anyway ...

**Comment by wenningerk** [comment]
> Do you mean on creation of the diff, when iterating over the local hash table, we start dropping unneeded entries from the remote hash table and even the local hash table already? Well, of course when merging the state, we'll still have to iterate over the remote hash table to go through the failing-pending logic anyway ...

Yes that is what I meant.
Still would be less entries to iterate over and logic would be much simpler - just move to local hash table or fail.

**Pull request merged**